distributed/utils_comm.py

```diff
@@ -22,52 +30,83 @@
 
 
 async def gather_from_workers(
-    who_has: Mapping[str, Collection[str]],
+    keys: Iterable[str],
+    who_has: Callable[
+        [list[str]],
+        Mapping[str, Collection[str]] | Awaitable[Mapping[str, Collection[str]]],
+    ],
```

Comment:
I'm not convinced this callback structure is the right approach here. I find this makes the entire mechanism even harder to reason about. Besides, this bypasses (for better or worse) missing/update_who_has mechanics on the worker.
I believe this kind of logic should be handled a layer further up the stack instead of throwing this all into this function.

Reply:
The method on the worker exclusively serves replicate, which we need to pen in to replace with AMM to begin with. The original method bypasses the whole worker missing/update_who_has system and I chose not to change that logic since it would be very complicated and ultimately throwaway.